core/src/main/scala/com.snowplowanalytics.stream/loader/executors/NsqSourceExecutor.scala

```diff
@@ -77,46 +80,72 @@ class NsqSourceExecutor(
     case Purpose.Bad      => new BadEventTransformer
   }
 
+  val executorService = Executors.newSingleThreadScheduledExecutor
+
   /**
    * Consumer will be started to wait new message.
    */
-  override def run(): Unit = {
-    val nsqCallback: NSQMessageCallback = new NSQMessageCallback {
-      val nsqBufferSize = nsq.buffer.recordLimit
-
-      override def message(msg: NSQMessage): Unit = {
-        val msgStr = new String(msg.getMessage(), UTF_8)
-        msgBuffer.synchronized {
-          val emitterInput = transformer.consumeLine(msgStr)
-          msgBuffer += emitterInput
-          msg.finished()
-
-          if (msgBuffer.size == nsqBufferSize) {
-            val rejectedRecords = emitter.attemptEmit(msgBuffer.toList)
-            emitter.fail(rejectedRecords.asJava)
-            msgBuffer.clear()
-          }
+  val nsqCallback: NSQMessageCallback = new NSQMessageCallback {
+    val nsqBufferSize = nsq.buffer.recordLimit
+
+    override def message(msg: NSQMessage): Unit = {
+      val bytes  = msg.getMessage
+      val msgStr = new String(bytes, UTF_8)
+      msgBuffer.synchronized {
+        val emitterInput = transformer.consumeLine(msgStr)
+        msgBuffer += emitterInput
+        msgBufferBytes += bytes.size
+        msg.finished()
+
+        if (msgBuffer.size == nsqBufferSize || msgBufferBytes > nsq.buffer.byteLimit) {
+          flush()
         }
       }
     }
+  }
 
-    val errorCallback = new NSQErrorCallback {
-      override def error(e: NSQException): Unit =
-        log.error(s"Exception while consuming topic ${nsq.streamName}", e)
+  def flush(): Unit = msgBuffer.synchronized {
+    if (msgBuffer.nonEmpty) {
+      val rejectedRecords = emitter.attemptEmit(msgBuffer.toList)
+      emitter.fail(rejectedRecords.asJava)
+      msgBuffer.clear()
+      msgBufferBytes = 0
     }
+  }
+
+  val errorCallback = new NSQErrorCallback {
+    override def error(e: NSQException): Unit =
+      log.error(s"Exception while consuming topic ${nsq.streamName}", e)
+  }
 
-    // use NSQLookupd
-    val lookup = new DefaultNSQLookup
-    lookup.addLookupAddress(nsq.nsqlookupdHost, nsq.nsqlookupdPort)
-    val consumer =
-      new NSQConsumer(
-        lookup,
-        nsq.streamName,
-        nsq.channelName,
-        nsqCallback,
-        new NSQConfig(),
-        errorCallback
-      )
+  // use NSQLookupd
+  val lookup = new DefaultNSQLookup
+  lookup.addLookupAddress(nsq.nsqlookupdHost, nsq.nsqlookupdPort)
+  val consumer =
+    new NSQConsumer(
+      lookup,
+      nsq.streamName,
+      nsq.channelName,
+      nsqCallback,
+      new NSQConfig(),
+      errorCallback
+    )
+
+  override def run(): Unit = {
+    val flusher = new Runnable {
+      def run(): Unit = flush()
+    }
+    executorService.scheduleWithFixedDelay(
+      flusher,
+      nsq.buffer.timeLimit,
+      nsq.buffer.timeLimit,
+      TimeUnit.MILLISECONDS
+    )
```

Comment:
With this design it will often be the case that we will re-flush although we have already flushed just a few ms ago.

To fix that the collector resets [lastFlushedTime](https://github.com/snowplow/stream-collector/blob/master/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L98) each time [it flushes](https://github.com/snowplow/stream-collector/blob/master/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L120) and then in the scheduled flush it [checks its value](https://github.com/snowplow/stream-collector/blob/master/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L136).